File: web/src/core/rakaatTracker.ts
import { BodyJoints } from "./geometry.js";
import { violationMessage } from "./movementValidator.js";
import { Posture } from "./posture.js";
import { PostureClassifier } from "./postureClassifier.js";
import { PostureSmoother } from "./postureSmoother.js";
import { PrayerTypeInference } from "./prayerTypeInference.js";
import { RakaatSession, makeSessionId } from "./rakaatSession.js";
import { RakaatStateMachine } from "./rakaatStateMachine.js";

export interface TrackerDeps {
  classifier?: PostureClassifier;
  smoother?: PostureSmoother;
  stateMachine?: RakaatStateMachine;
  inference?: PrayerTypeInference;
}

/** Top-level pipeline the UI drives: classify → smooth → state machine. Feed it
 *  joints + timestamp per frame; read live state for the overlay and
 *  makeSession() once complete. Holds no frames. */
export class RakaatTracker {
  private classifier: PostureClassifier;
  private smoother: PostureSmoother;
  private stateMachine: RakaatStateMachine;
  private inference: PrayerTypeInference;

  constructor(deps: TrackerDeps = {}) {
    this.classifier = deps.classifier ?? new PostureClassifier();
    this.smoother = deps.smoother ?? new PostureSmoother();
    this.stateMachine = deps.stateMachine ?? new RakaatStateMachine();
    this.inference = deps.inference ?? new PrayerTypeInference();
  }

  get livePosture(): Posture {
    return this.smoother.current;
  }
  get rakaatCount(): number {
    return this.stateMachine.rakaatCount;
  }
  get isComplete(): boolean {
    return this.stateMachine.isComplete;
  }

  /** Process one frame. `t` is a monotonic timestamp in seconds. */
  process(joints: BodyJoints, t: number): Posture {
    const raw = this.classifier.classify(joints);
    const committed = this.smoother.feed(raw);
    this.stateMachine.update(committed, t);
    return committed;
  }

  /** End the session early (user taps stop). */
  stop(t: number): void {
    this.stateMachine.finish(t);
  }

  /** Build the persisted session result. Call after isComplete (or stop). */
  makeSession(date: Date = new Date()): RakaatSession {
    return {
      id: makeSessionId(),
      timestamp: date.getTime(),
      prayer: this.inference.inferAt(this.stateMachine.rakaatCount, date),
      rakaat: this.stateMachine.rakaatCount,
      violations: this.stateMachine.violations.map(violationMessage),
    };
  }

  reset(): void {
    this.smoother.reset();
    this.stateMachine.reset();
  }
}

File: Sources/RakaatCore/RakaatTracker.swift
import Foundation

/// Top-level pipeline the iOS app drives. Feed it joints + timestamp per frame;
/// read `livePosture` / `rakaatCount` for the overlay, and `makeSession(...)`
/// once `isComplete` to get the persisted result.
///
///     classify → smooth → state machine (+ validator)
public final class RakaatTracker {
    private let classifier: PostureClassifier
    private let smoother: PostureSmoother
    private let stateMachine: RakaatStateMachine
    private let inference: PrayerTypeInference

    public init(
        classifier: PostureClassifier = PostureClassifier(),
        smoother: PostureSmoother = PostureSmoother(),
        stateMachine: RakaatStateMachine = RakaatStateMachine(),
        inference: PrayerTypeInference = PrayerTypeInference()
    ) {
        self.classifier = classifier
        self.smoother = smoother
        self.stateMachine = stateMachine
        self.inference = inference
    }

    public var livePosture: Posture { smoother.current }
    public var rakaatCount: Int { stateMachine.rakaatCount }
    public var isComplete: Bool { stateMachine.isComplete }

    /// Process one frame. `t` is a monotonic timestamp in seconds.
    @discardableResult
    public func process(joints: BodyJoints, t: Double) -> Posture {
        let raw = classifier.classify(joints)
        let committed = smoother.feed(raw)
        stateMachine.update(posture: committed, t: t)
        return committed
    }

    /// End the session early (user taps stop).
    public func stop(at t: Double) {
        stateMachine.finish(at: t)
    }

    /// Build the persisted session result. Call after `isComplete` (or stop).
    public func makeSession(at date: Date = Date(), calendar: Calendar = .current) -> RakaatSession {
        let prayer = inference.infer(rakaat: stateMachine.rakaatCount, at: date, calendar: calendar)
        return RakaatSession(
            timestamp: date,
            prayer: prayer,
            rakaat: stateMachine.rakaatCount,
            violations: stateMachine.violations.map(\.message)
        )
    }

    public func reset() {
        smoother.reset()
        stateMachine.reset()
    }
}
